db/db.ts

```diff
@@ -448,3 +453,103 @@ export const getNonGrapherExplorerViewCount = (
             AND grapherId IS NULL`
     ).then((res) => res?.count ?? 0)
 }
+
+/**
+ * 1. Fetch all records in tag_graph, isTopic = true when there is a published TP/LTP/Article with the same slug as the tag
+ * 2. Group tags by their parentId
+ * 3. Return the flat tag graph along with a __rootId property so that the UI knows which record is the root node
+ */
+export async function getFlatTagGraph(knex: KnexReadonlyTransaction): Promise<
+    FlatTagGraph & {
+        __rootId: number
+    }
+> {
+    const tagGraphByParentId = await knexRaw<FlatTagGraphNode>(
+        knex,
+        `-- sql
+        SELECT
+            tg.parentId,
+            tg.childId,
+            tg.weight,
+            t.name,
+            t.slug,
+            IFNULL((p.type IN (:types) AND p.published = 1 AND t.slug IS NOT NULL), FALSE) AS isTopic
+        FROM
+            tag_graph tg
+        LEFT JOIN tags t ON
+            tg.childId = t.id
+        LEFT JOIN posts_gdocs p on
+            t.slug = p.slug
```

Comment:
Energy was duplicated for me - if you move the logic from isTopic into the join condition then the unpublished energy gdoc does not lead to a second row for the tag and then you don't have that problem
```suggestion
            p.slug is not null AS isTopic
        FROM
            tag_graph tg
        LEFT JOIN tags t ON
            tg.childId = t.id
        LEFT JOIN posts_gdocs p on
            t.slug = p.slug AND p.published = 1 and p.type IN (:types)
```